### ISSUE TITLE
Full Demo application freezes when using the DTWAIN64u.dll (HP 7000 S3)

### ISSUE THREAD

**Issue by sean-neeley** [dynarithmic/twain_library-csharp_demo#1]
[FullDemo.zip](https://github.com/dynarithmic/twain_library-csharp_demo/files/10513287/FullDemo.zip)

We are using an HP Scanjet 7000 s3.  The attached sample program works great when running as "Debug - Any CPU", which uses DTWAIN32u.dll.  However if you take the same project and run as "Debug - x64" which uses DTWAIN64u.dll the application freezes during the HP "Processing Images" dialog.  We our trying to build a 64bit application and this is a show stopper for us.

<img width="303" alt="Untitled" src="https://user-images.githubusercontent.com/80922869/214949350-157ddaea-9bd7-4284-824e-23ac5c2b3bb2.png">

To recreate, run the attached project as Debug - x64.  
Put about 5 sheets of paper in the document feeder.
Select the source, then Acquire Test -> Acquire Buffered from the menu
Move the mouse outside of the application window right away (not sure if this is really needed)
The HP scanner dialog will show it is acquiring the images, and feeds the paper through.
Next the HP scanner dialog will show it is Processing the images, but will hang.
If you run the same steps above with the "Debug - Any CPU" and DTWAIN32u.dll, it will not hang in this spot and will populate the document tree view with an entry for each page scanned.

Unfortunately we do not have any other model scanners with 64bit TWAIN drivers to test this.  It could be only an issue that happens with the HP drivers, but I'm not sure.  I believe you said you had an HP scanner.

Thank you

**Comment by dynarithmic** [comment]
What happens if you acquire using native mode instead of buffered?  

I will see if there is an issue with the buffered mode if the native mode goes through ok.

**Comment by dynarithmic** [comment]
It looks like the native mode should work for you, since buffered mode is a little more complex in terms of memory management.

Do you have another application, preferably one *not* written by HP, that accesses the scanner and does transfers using the 64-bit driver?  If you do, and you can control whether you can acquire using native or buffered, that would be a good test to see if there is an underlying problem with DTWAIN and 64-bit buffered transfers.

**Comment by sean-neeley** [comment]
Thank you for the quick reply.  We will give this a try and report back next week.

**Comment by dynarithmic** [comment]
I will be making a few changes for the buffered mode.  When the changes are ready, I will make them available in the [development branch of the twain_library_source repository](https://github.com/dynarithmic/twain_library_source/tree/development/binaries/64bit).  You can test them to see if any difference is made when using the HP.  

I will let you know when the new DLL's become available.

**Comment by sean-neeley** [comment]
OK Thanks Paul!

**Comment by dynarithmic** [comment]
I just updated the [development branch](https://github.com/dynarithmic/twain_library_source/tree/development/binaries/64bit) with version 5.3.0.1 of DTWAIN.   The release_libraries.zip contains the latest version.  

The changes are to address a potential issue with the buffered transfer calling the TWAIN system to query when there are more pages to transfer.   Possibly this is the issue you encountered, but not sure until you test the new version.  

This was tested against a Plustek scanner using 64-bit drivers, and the sample TWAIN Working Group 64-bit driver.   Both scanners work with 5.3.0.0, but I found issues without showing the scanner user-interface in buffered mode, and the TRUE/FALSE "close source when UI is closed" parameter to DTWAIN_AcquireBufferedEx() that could affect things.

**Comment by sean-neeley** [comment]
OK, I download your release_libraries.zip and tried the latest 64bit DLL.  Unfortunately the test application still freezes.  I tried the following:
- Switched code to AcquireNativeEx.  No luck
- Tried with the close source option both TRUE and FALSE. No luck
- Removed the callback function from the code.  No luck

Application still freezes on the "Processing Image(s)" dialog.  If I pause the debugger, it pauses on the AcquireBufferedEx() line.

I downloaded Twacker_64 to test.  I set it to hide UI and use Memory for the acquire.  
<img width="327" alt="image" src="https://user-images.githubusercontent.com/80922869/215859441-8ea171ee-bd25-4da1-873f-86acd6d1d8da.png">
It works without hanging and "Processing Image(s)" appears and goes away as expected.

Is there anything I can try next?  Can I create any logs to help you?

**Comment by dynarithmic** [comment]
Are you acquiring with or without the UI when you're using DTWAIN?  If you are acquiring without the UI, please try this with the UI being shown.

Did you also try the DTWDEMO program?  Twacker is written in C, just like the DTWDEMO program.  Make sure the issue has nothing to do with .NET.

**Comment by dynarithmic** [comment]
As to logs, you can use DTWAIN_SetTwainLog and turn on all logging and log to a file (the DTWDEMO program has such an option).  

One difference in Twacker and DTWAIN is that Twacker does the minimal setup to scan an image -- it really isn't designed to exercise the scanner's capabilities as DTWAIN attempts to do.  Quite possibility the querying / setting of a capability is exposing something in the HP scanner that shouldn't occur.  It wouldn't be the first time this has happened (the Samsung Galaxy TWAIN driver had several issues where it wasn't following the TWAIN specifications).

**Comment by dynarithmic** [comment]
Also, if you are familiar with C++ and debugging on the native side of the app, you could debug into AcquireBufferEx().  To do that, you should use the dtwain32ud.cs file, make sure that dtwain32ud.dll is available, and try to debug into the native code.  

The debug dlls are found [here](https://github.com/dynarithmic/twain_library_source/blob/main/binaries/64bit/debug_libraries.zip), and the source [here](https://github.com/dynarithmic/twain_library_source/tree/main/source).  

The source file that may be of interest is [ctltr031.cpp](https://github.com/dynarithmic/twain_library_source/blob/main/source/cpp/ctltr031.cpp), where a bulk of the buffered transfer logic is being done.  Granted it is a lot of code, but I think you could see if the hanging of the app is occurring within that file.

**Comment by sean-neeley** [comment]
Enabling UI doesn't help.  I created logs for it working with the 32bit DLL and it not working with the 64bit DLL.  showUI was false for these logs.

[DTWAIN32.LOG](https://github.com/dynarithmic/twain_library-csharp_demo/files/10551690/DTWAIN32.LOG)
[DTWAIN64.LOG](https://github.com/dynarithmic/twain_library-csharp_demo/files/10551692/DTWAIN64.LOG)

I'm slightly familiar with C++ and could probably debug it.  However I'm not sure of how to debug ctlr031.cpp from my C# project.  Or would I locate a C++ DTWAIN demo app and debug in there?

**Comment by dynarithmic** [comment]
Did you try the DTWDEMO.EXE program?  Still not sure if this is a C# / .NET issue or a general issue.

**Comment by sean-neeley** [comment]
I downloaded DTWDEMO.EXE, but it is 32bit.  It works.  

<img width="242" alt="image" src="https://user-images.githubusercontent.com/80922869/215904352-5fa71415-6482-4373-ab8b-6e270d61402d.png">

Is there a 64bit version somewhere?

**Comment by dynarithmic** [comment]
I placed the demo programs in the source repo.  You can get it [here](https://github.com/dynarithmic/twain_library_source/tree/main/binaries/64bit).

**Comment by dynarithmic** [comment]
Looking at the DTWAIN64.LOG, there is a failure of setting the feeder.  If you search the log file for TWCC_CAPBADOPERATION, you should see it around line 34012:

`    ConditionCode=14  TWCC_CAPBADOPERATION`

Preceeding this, an attempt to set the feeder was done, but resulted in failure.  You can see the previous attempt to set the feeder, and the failure detected around line 33896:

`[2023-01-31 16:32:16] Output: return code from DSM=1 (TWRC_FAILURE)`

and the TWCC_CAPBADOPERATION is what was returned when querying as to why the setting of the feeder failed.  Note that in the 32-bit log, around the same sequence of calls, there is no failure in setting the feeder.  

As demonstrated by the logs, there is a difference in the way the HP 32-bit driver and 64-bit driver behave, and it is not a straight port of the driver code from 32-bit to 64-bit.  I also saw that that the 32-bit driver has 117 capabilities, while the 64-bit driver has 118 capabilities.  There is a ICAP_IMAGEMERGE capability that doesn't exist in the 32-bit version.  It is strange why the same hardware would have differences in the driver like this.

Unfortunately, I don't have this scanner model.  So debugging the issue would be a very slow process, unless I get the scanner here and test it.  Probably wouldn't take me long to figure out a workaround, but it is very difficult without the hardware.

**Comment by dynarithmic** [comment]
To debug, you could load the dtwain_5_vs2019.sln into Visual Studio, and attach to the running C# process.  Then you should be able to set a breakpoint in the C++ code (DTWAIN_AcquireNativeEx() found in cltacquirenative.cpp would be a starting point.

**Comment by dynarithmic** [comment]
The new demo programs include the changes that were made in 5.3.0.1, so the issue I identified a few days ago is accounted for if you get the demo programs at the previous link.

The [twain_library repository](https://github.com/dynarithmic/twain_library/tree/master/binaries/64bit) is still on 5.3.0.0, so the demo programs located there in the zip file won't have these changes.

**Comment by sean-neeley** [comment]
I'm having trouble running the DTWDEMO64.EXE.  I get this error:

<img width="380" alt="image" src="https://user-images.githubusercontent.com/80922869/216097928-00d48fcb-e46f-4ae0-8fc3-4c3a3135585f.png">

I downloaded and installed the MS Visual Studio x64 redistributable and rebooted, but that did not help.  I've never seen that error before.  Any ideas what I need to run it?

**Comment by dynarithmic** [comment]
I updated the development branch with the lastest 64-bit demo.  Please make sure that the DLLs and EXE's used are from [this archive](https://github.com/dynarithmic/twain_library_source/tree/development/binaries/64bit) in the 64bit_FullDemo.zip file.

Also, I took another look at the 32-bit DTWAIN log file you produced, and the version of DTWAIN that is running for 32-bit was 5.2.0.8.  Since this is an older version, did you make sure that version 5.3.0.0 for 32-bit also runs correctly?

**Comment by sean-neeley** [comment]
OK, thanks for the help.  I got DTWDEMO64.exe to run using your 64bit_FullDemo.zip.  It has the hanging issue.  One thing I noticed when testing it.  After it hangs, if I right click on the application in the Windows taskbar and choose "Close Application", suddenly it unfreezes and is able to continue operating OK.  I'm attaching a log of this.
[DTWAIN64new.log](https://github.com/dynarithmic/twain_library-csharp_demo/files/10561640/DTWAIN64new.log)
We noticed something similar in our C# app, where we would move the mouse or simulate a mouse cursor move and it would unfreeze.  I tried this mouse hack in the FullDemo app and it didn't do anything though.

Regarding the 32-bit DTWAIN log.  I created a new log with the 5.3.0.0 for 32-bit.  It ran fine, although it seemed a lot slower than the 64-bit application when selecting a source.  I do have more 32bit sources on my test system though.  I'm including the log below.  It is from the C# Full Demo.
[DTWAIN32new.log](https://github.com/dynarithmic/twain_library-csharp_demo/files/10561655/DTWAIN32new.log)

Last, I grabbed the DTWAIN dll source and tried to compile it so I could debug it.  However the project seems to require some library called "boost" that I do not have and is not part of the project or visual studio.  Do I need to download from boost.org?

**Comment by dynarithmic** [comment]
1) Looking at the 64-bit log, it is the setting of the feeder using the CAP_FEEDERENABLED capability that seems to be glitching things.  Then as you mentioned, it looks like the scan goes through afterwards, at least the scanner responds.   I will try to see if there is anything strange that can occur if the setting of the feeder fails (it shouldn't fail, as the 32-bit version has no issues doing this).

2) To rebuild the source, you will need the boost library, as [described here](https://github.com/dynarithmic/twain_library_source/blob/main/README.md).  

You also will need to set the BOOST_INCLUDE_DIR environment variable before starting Visual Studio.   The 7-zip and DirHash programs are not important in building the library (in your case) -- they just provide a way to package up and compute the hash code of the generated output files.

**Comment by dynarithmic** [comment]
I am speculating that it has something to do with the message processing and the UI that HP uses for 64-bit.  The strange behavior of the scan going through afterwards is unusual.

I may be able to get my hands on a HP 5000 S4, which may use the same base driver as the 7000 S3.   I went to the HP website to see what driver the 5000 uses, and the version number of the latest drivers for both 5000 S4 and 7000 s3 are "41.x".  

If my speculation is correct, that the driver's base code is the same, then I may be able to duplicate the issue here.  However, I won't really be able to do anything until I can confirm I can get the scanner, and have it in-house.   That may take more than a week.   

The only other option (and this has been done before, but this was when DTWAIN was commercial) is for the user to send the scanner to us and we fix the issue.  It usually takes a very short time, since we are familiar with the code and know what to look out for and if need be, write any workarounds for bugs that may be in the scanner driver.  Not sure if you want to go this route, but let me know.

**Comment by sean-neeley** [comment]
OK thank you.  I am going to see how far I can get with debugging myself.  If I'm unable to locate the problem, we should be able to send you a spare 7000 s3 that we have.  Are you located in the united states?

**Comment by dynarithmic** [comment]
Yes, I'm here in the US.  Just let me know when/if you will send the scanner over.  

In the meantime I may have to get this model or one based on it in-house on a permanent basis.  I have for 64-bit scanning a Plustek and Fujitsu scanner for ADF/duplex testing, but the HP scanner I have doesn't have a 64-bit driver, so can't use it for 64-bit access.

**Comment by dynarithmic** [comment]
Good news:

I landed getting an HP 7000 scanner.  I will probably also get the HP 5000 in-house.   I should get it by the end of this week, and will probably be ready to test sometime next weekend.

**Comment by sean-neeley** [comment]
Ok, that is awesome.  I'm still working on getting the DTWAIN dll to compile.  You would definitely be able to figure out the issue a lot quicker than I could.  I appreciate the help.

**Comment by dynarithmic** [comment]
Yes, hopefully I can duplicate the issue, and the fix will not be that difficult.

To get the DLL built, it is basically 4 steps:

1) Get the [boost library for your Visual Studio platform setting](https://sourceforge.net/projects/boost/files/boost-binaries/1.81.0/).  The default platform is 2015 (v140), so you can install the libraries (32 and 64 bit) that have file names ending with "14.0-32" and "14.0-64".  

If you happen to not have the v140 runtime platform installed, then you should get the other 32 and 64 bit boost libraries that fit the platform you have (for example, Visual Studio 2019 always asks if you want to upgrade to v142 when you load a project with 2015 components -- you don't have to do this, but just in case you do...).  

The platform for v142 is 14.2, for v141 it is 14.1, etc.  So download the boost files that are named appropriately.

When installing boost, make sure that the 64-bit and 32-bit versions are installed in the same directory (the installation will also add the header files).  So if you installed boost in **C:\boostlibs**, then you should see a **boost** directory under C:\boostlibs, and a **lib32-msvc-14.0** and a **lib64-msvc-14.0" directory (assuming you are using the v140 platform).  

2) Set the environment variables:

Using the example above, you can use the command-line, or use the Settings dialog in Windows OS.  Here is the command:

SET BOOST_INCLUDE_DIR=c:\boostlibs
SET BOOST_LIBRARY_DIR_V140_32=c:\boostlibs\lib32-msvc-14.0\
SET BOOST_LIBRARY_DIR_V140_64=c:\boostlibs\lib64-msvc-14.0\

3) Start Visual Studio 2019 (or 2022) and load the dtwain_5_3_vs2019.sln as the solution file.    You may be prompted to upgrade to 142 platform.  If you have v140 installed, then do not upgrade, if you don't have it installed and need to upgrade to match what platform you do have installed, then make sure your boost files match the platform you are upgrading to (see step1).   

4) On the Visual Studio main menu, choose "Build -> Batch Build" and select all of the projects named DTWAIN_DLL and DTWDEMO.  Then choose "Build" from that same dialog.

That's about it.  Any errors you get when building will be for the 7-zip utility and the DirHash utility not being found, but that isn't going to affect the build of the dlls.  Once built, the DLLs will be in a **binaries** directory under the **source** directory.

Hope that helps.

**Comment by sean-neeley** [comment]
OK, I'm using platform v143.  I compiled boost 1.81.  It has a little different directory structure.  All of the libraries seem to be in the folder c:\boost_1_81_0\staged\lib\.  I opened the DTWAIN project and upgraded to 143 when prompted.  I then did the batch build and selected all.  It took an hour or so and then I end up with these two errors:

<img width="1099" alt="image" src="https://user-images.githubusercontent.com/80922869/217035010-bfc42ffb-0f5d-45e6-a0ab-0377f899b1c3.png">

These are the closest matching library files on disk:

<img width="805" alt="image" src="https://user-images.githubusercontent.com/80922869/217034938-a5e169d9-f1f3-48d6-8246-273a28c00425.png">

It looks like the only difference in the filename is "sgd" vs "gd" and I'm not sure why.  Could I just rename it on disk?  Or should I change it somewhere in the project?  Or should I have done something different when building boost?

Thanks

**Comment by dynarithmic** [comment]
I am currently using boost 1.80.  If the other DLL's built without error, possibly this is missing from the list of binaries that were built for version 1.81.  On my system, I have this file for platform v143:

libboost_filesystem-vc143-mt-sgd-x32-1_80.lib

I haven't tried to build 1.81 and VS 2022, so I am not aware of this issue.  I have used VS 2019 and boost 1.80 without an issue.  I will download it to see if that file is missing for VS 2022.

In the meantime, you can use the 64-bit DLL's (if they built correctly), or if you want to try to build everything again, you can try this:

Since you have the platform set to V143, what you could do is go to each of the individual project files in the Solution (go to the Solution Explorer, and expand everything), and set the compiler to 

**ISO C++17 Standard (/std:c++17)**

To get to this setting, highlight the project in solution explorer, right click, select "Properties".  
Then select **C/C++ -> Language -> C++ Language Standard**

Then rebuild.  What this will do is remove the dependency of boost filesystem from DTWAIN.  The reason why there is a dependency right now is that the v140 and v141 platforms are not C++17 compliant, and that is minimum required to have built-in filesystem support in C++.

You will need to do this for each of the following projects:

DTWDEMO
cryplib
all of the FreeImage projects (FreeImageLib, LibJPEG, LibJXR, etc.)
DTWAIN_DLL

Once you change the compiler compliancy in each one, then you can do a rebuild.

**Comment by dynarithmic** [comment]
I just downloaded boost 1.81 for platform v143, and the file is included in the lib folder:

```
 Directory of C:\boost_1_81_0\64\lib64-msvc-14.3

12/08/2022  05:00 PM         8,069,770 libboost_filesystem-vc143-mt-sgd-x64-1_81.lib

```
This is only for 64-bit.  I will check the 32-bit version and see what happens on a rebuild using VS 2022 .

**Comment by sean-neeley** [comment]
OK, so all of my boost library files have "-gd-" or "-mt-" in them.  I searched for info on this and the "-sgd-" is for a statically linked build of the library.  So somehow I missed an option to build statically linked libraries.  I'll look into it tomorrow.  I can also try your idea to skip using boost.  Thanks again.

UPDATE: I think I need to use 'b2 link=static' when building the boost libs

**Comment by dynarithmic** [comment]
ok.  I only use the boost binaries for Windows, and not need to rebuild them.  But nothing wrong in attempting to build the boost libs yourself.  The last time I had to build them was for a Linux port of DTWAIN (which is on hold right now).

Also, I was able to build using VS 2022 using v143 and C++17.   However there is a coding change in some third-party imaging code you will need to make.  You can pick up the latest source in the [development/source repo](https://github.com/dynarithmic/twain_library_source/tree/development/source).    

If you are familiar with C++ and smart pointers, the issue is that the old imaging code uses **std::auto_ptr**, and this class was removed from C++17, so the C++17 build failed.  This had to be changed to unique_ptr.

**Comment by sean-neeley** [comment]
Thanks I did run into that compiler issue and didn't know what to do.  I'll try the latest source repo.

**Comment by sean-neeley** [comment]
Hi, I tried the new source repo and deleted my boost folder and changed every project to use ISO C++17 Standard (/std:c++17).  However now I get compiler errors for missing boost includes.  Do I need to keep around the boost folder for the includes?  Example includes that are flagged:

```
#include <boost/tokenizer.hpp>
#include <boost/algorithm/string/join.hpp>
#include <boost/algorithm/string.hpp>
#include <boost/uuid/uuid.hpp>
#include <boost/uuid/uuid_generators.hpp>
#include <boost/uuid/uuid_io.hpp>
#include <boost/format.hpp>
```

**Comment by dynarithmic** [comment]
Yes, you need to keep the boost include directories.

**Comment by dynarithmic** [comment]
OK, I received the 7000 S3 scanner today and did some rudimentary testing.  Here are my observations:

1) I used DTWDEMO64UD.exe, the debug version of the DTWDEMO app for 64-bit.  I tested just a single page.

2) When opening the HP driver (DTWAIN_OpenSource), the driver will throw an exception with the following error:
`Exception thrown at 0x00007FF8EFBCCD29 in DTWDEMO64UD.exe: Microsoft C++ exception: HP::Pdsmq::Common::FileNotFoundException at memory location 0x000000EE487E70E0.
ERROR: Fail creating Settings class: No factory settings file found
`
This exception doesn't stop the application, just something that I noticed.  So I ignored all of those exceptions and continued.

3) In the DTWDEMO app, I chose to scan using no user interface, made sure the document feeder was used, and buffered transfer.  The scan for me went through, but there is something that annoys me with the HP driver, and possibly this is what is happening on your end:

After obtaining all of the transferred strips, the HP Scan Utility application opens up.  For me it opened, but I have no idea what would happen on a system where the HP app wasn't installed, had trouble starting, etc.  I don't know a workaround to turn off the way this woks, but I can imagine the issues if after the scan, something hinders that application from starting.

What's troublesome is that while this is happening, the DTWDEMO application is in the TWRC_XFERDONE message processing.  If you are able to get a debug build to run, you will see this in ctltr031.cpp, around line 245 (you can put a breakpoint there).  That is where I verified that all the strips were transferred, **and** the HP Utility app tries to start.  

4) After this, more exceptions are thrown by the HP driver, similar to the initial one where an xml file cannot be located.  The application ignores them, and the HP driver closes.

So I didn't get the hanging issue, but I certainly could see why an application will have issues with the driver if one or more of the utility apps don't start, or not installed properly.

**Comment by sean-neeley** [comment]
Thanks Paul.  I'll be back in the office on Tuesday and that is when I can test again.  I can let you know what HP software packages we have installed.  I have seen that HP dialog appear before in the DTWAIN demo app, but I don't remember under what circumstances.  It never appears in our own C# DTWAIN application, but like you said that could be related to the freezing.

I'm still struggling getting the DTWAIN dll compiled.  Without compiling boost I get errors about missing boost related files.  When I do compile boost, the DTWAIN project is looking for boost libraries with different names than what I have on disk.  I can give more details on Tuesday.

**Comment by sean-neeley** [comment]
OK, Regarding the DTWAIN library build, I changed the compiler setting to /std:c++17 and have the boost folder in my include path.  It seems to batch build OK, but I get these 2 link erors:

<img width="1109" alt="image" src="https://user-images.githubusercontent.com/80922869/218834124-fb3b76fd-d262-44b3-9e7c-316d2cdcdc2e.png">

Regarding the HP scanner freezing...  For the software we install the following packages from the HP website, in this order:
- HP ScanJet Flow 7000 S3 Full Feature Software and Driver
- HP ScanJet Enterprise Flow 7000 s3 ISIS Driver and Utilities
- KOFAX VRS 5.0
- HP ScanJet Flow 7000 s3 Basic Driver

We use KOFAX and the ISIS driver for an older build of our application that doesn't use DTWAIN, so that is why they are installed.

When using the DTWDEMO app, the HP Scanner Tools Utility does open the first time I scan, which is annoying  For subsequent scans and runs of the application I do not see it.  I found where the HP Scanner Tools "bin" folder is located and renamed it, just to see if that would stop the freeze up, however that did not help.  When using the modified DTWDEMO app that I gave you at the beginning of this bug report, the app freezes on the "Processing Image(s)" when debugging in visual studio.  However if I click on the Visual Studio icon in the taskbar, the action of switching focus away from the DTWDEMO window somehow makes it unfreeze.

**Comment by dynarithmic** [comment]
What options are you using to scan the image in DTWDEMO?  Are you using the user interface?  Is it buffered or native?  

As to the linker error, what are the environment variable **BOOST_LIBRARY_DIR_V143_64** and **BOOST_LIBRARY_DIR_V143_32** set to?  Are the libraries mentioned in the error found in those directories specified by the environment variables?  The ones on my system are set to:

```
BOOST_LIBRARY_DIR_V143_32=c:\boost_1_81_0\lib32-msvc-14.3\
BOOST_LIBRARY_DIR_V143_64=c:\boost_1_81_0\lib64-msvc-14.3\

```

**Comment by sean-neeley** [comment]
For DTWDEMO, I'm using AcquiredBufferedEx and No UI.  If you use the FullDemo.zip in my first post, it is a modified version of the demo that is hardcoded to use AcquiredBufferedEx and No UI.  It is what I have been using to recreate the freezing problem.

I don't have any libraries anywhere in C:\boost_1_81_0, because I did not compile boost.  I had trouble in the past getting it to build the static version of the boost libraries.  Maybe I understood wrong, but I thought you said that if I use the /std:c++17 in my DTWAIN project, I would not need the boost libraries, only the header files.  I'm trying to compile the DTWAIN DLL without the boost libs.

Thank you

**Comment by dynarithmic** [comment]
I think the compilation issue is due to Visual C++ not setting all of the flags when C++17 is selected.  You should set the **/Zc:__cplusplus** option.  You can do it by opening up the Properties for the project, then:

**Configuration Properties -> C/C++ -> Command Line**

and then in the Additional Options box at the bottom of the dialog, add:

**/Zc:__cplusplus**

Recompile.  See if that solves the issue.

Also, there are still some components of boost that are required in library form, such as the dll component.  However the filesystem component shouldn't be necessary once the C++17 compilation is successful.

**Comment by dynarithmic** [comment]
I just ran with the latest version of DTWAIN (5.3.0.1), 64-bit.  Except for the exceptions being thrown, I had no issues with the C# project scanning with no UI, buffered mode (I tried a single page -- maybe you should try this also).

I did see some issues with the C# project, but these are probably minor:

1) If there is no paper in the feeder, and the source UI is not displayed, you will get a condition code error of 29, meaning there is no paper in the feeder.  This might be an issue, since there is a C++ wrapper class that uses a timer, where the user can wait for a certain number of seconds before placing paper in the feeder.  

The other issue is that the C# code doesn't test for this error, and hangs on the spot where it detects that no images have been captured.  A call to DTWAIN_GetLastErrorString(DTWAIN_GetLastError)) should show that the paper has not been inserted into the feeder tray.

2) The HP scan utility pops up for me when all the strips have been transferred, so I'm still not so sure about the way this is supposed to work on the HP side.  

3) I am still concerned about the exceptions being thrown by the driver because of some missing configuration files.  I may have to see if there is any way to just log those exceptions and not have them interrupt the flow of the program.

But other than those issues, I didn't have a problem once the paper is placed in the feeder, the page is read (the paper basically flies out of this model of scanner -- not used to that, lol), and the image window displayed what was captured.  

I used the latest driver version from HP for the 7000 S3.  Here is the information, when logged by DTWAIN:

```cpp
    Id=1
    Version Number=2.1
    Version Language=2
    Version Country=1
    Version Info=2.1
    ProtocolMajor=2
    ProtocolMinor=1
    SupportedGroups= DG_CONTROL, DG_IMAGE, DF_DS2
    Manufacturer=HP Inc.
    Product Family=HP
    Product Name=HP TWAIN USB
```

Maybe try the following:

Turn off the callback in the C# program.  That is done by commenting out

```cpp
        TwainAPI.DTwainCallback64 twainCallback = TwainCallbackProc;
        TwainAPI.DTWAIN_SetCallback64(twainCallback, 0);
```
Maybe the callback is causing the issue, not sure.  I am not a C# expert, but I've always had some trouble with getting the callback mechanism to work properly, more probably because of my non-expert knowledge of C#.

**Comment by dynarithmic** [comment]
Also, have you tried to scan using another PC?  Maybe there is something odd about your configuration.

**Comment by sean-neeley** [comment]
Yep, I have that same driver version.
```
TW_MEMREF is TW_IDENTITY:
Decoded TW_MEMREF:
{
    Id=1
    Version Number=2.1
    Version Language=2
    Version Country=1
    Version Info=2.1
    ProtocolMajor=2
    ProtocolMinor=1
    SupportedGroups= DG_CONTROL, DG_IMAGE, DF_DS2
    Manufacturer=HP Inc.
    Product Family=HP
    Product Name=HP TWAIN USB
}
```

When I test, I always use 2 or more sheets of paper.  I haven't checked if that really matters or not.

I tried commenting out the callback and not using, because I had the same thought.  It did not help unfortunately.

We also tried from another workstation and with another HP 7000 s3 scanner and it has the same issue.  I can try from a third workstation, where I only install the HP basic driver.

Can you tell me how you built boost?  I downloaded the latest version, unpacked it to C:\boost_1_81_0, then ran bootstrap.bat, followed by "./b2".  I seem to have a different folder structure and library names from you.  Thanks.

**Comment by dynarithmic** [comment]
For boost, I go to [here](https://sourceforge.net/projects/boost/files/boost-binaries/1.81.0/) and get the relevant binaries.  

The installation automatically puts the headers and libraries in the correct folders.  So for the Visual C++ family of compilers, there is no need to do a boostrap build, as the libraries are pre-built.

**Comment by dynarithmic** [comment]
Not sure how to get further with this, except to point out that the HP driver throws exceptions and will start the relevant applications.  

If you can get boost to build, the way you debug the full demo is to 

1) In the Solution Explorer, select the DTWAIN_DLL project, right click, and make it the Startup Project by select **Set as Startup Project**
2) Make sure the newly built DTWAIN DLL is the one your C# code is using.
3) In the Project properties:
**Configuration Properties -> Debugging** and then Command, you enter the full path of the FullDemo.exe executable.

Then start debugging (F5/F10).  What will happen is that FullDemo.exe will start, and then you can set breakpoints in the C++ code.  You can start with ctlacquirebuffered.cpp, where the DTWAIN_AcquireFileBufferedEx starts.  The file responsible for the buffered transfer from the device to DTWAIN is **ctltr031.cpp**, where you see the **switch** statement with the 

`case TWRC_XFERDONE:`

around line 244 of that file.  That code is executed when all of the strips have been transferred.

Note that starting the application from the DLL project is the only foolproof way of getting the C++ code to be debugged under C#.  I tried to debug the code from the FullDemo C# project, but even selecting to debug the native code resulted in not being able to get to the C++ layer.

**Comment by dynarithmic** [comment]
To verify, here is a snapshot of the program after successfully scanning a page:

![Clip1](https://user-images.githubusercontent.com/1283116/218934995-b6ac4ddc-5f4f-44bb-8aae-35d72ce8536a.jpg)

**Comment by sean-neeley** [comment]
OK, I was able to install boost using the exe and was able to get the 64bit dtwainu.dll to compile.  I missed where you posted a link to the exe in an earlier reply, so I apologize for that.  I was able to set a breakpoint in ctlacquirebuffered.cpp where the acquired buffered logic starts.  When stepping through the code line by line from that point on, it doesn't freeze.  However, if I test again and hit the breakpoint in the acquired buffered logic and hit the "play" button to continue execution of the program instead of stepping, it does freeze.  I'm not sure how to tell where it is freezing.  I hit the "pause" button in the debugger and it shows these threads:

<img width="1164" alt="image" src="https://user-images.githubusercontent.com/80922869/219137963-1d53af00-8909-48b0-ab46-b0cdfdfd15ef.png">

Clicking between the application window and the visual studio window seems to make it unfreeze.  Any advice for debugging and troubleshooting this sort of thing?

**Comment by dynarithmic** [comment]
I would look at the call stack to see what series of functions are being called.

The hpsxTwain.ds is the HP driver DLL that the TWAIN system is using to communicate with.  Other than that, there isn't really anything directly related to DTWAIN in those threads that I can tell.  If anything, possibly a call to the TWAINDSM.DLL (the 64-bit data source manager) is being hung on a call, and it is stuck in the HP driver code somewhere.

To verify this, the **ctltwmgr.cpp** file contains a function called **CallDSMEntryProc**.  In there, there should be a lines of code like this:
```cpp
    try
    {
        retcode = (*m_lpDSMEntry)( pOrigin, pDest, nDG, nDAT, nMSG, pData ); 
    }
    catch(...)
```
If the call stack points to that line of code (the **retcode = ** ), then this means that the driver or the TWAIN system is hanging on a certain triplet call.  To get the triplet that was sent to TWAIN, the values of **nDG, nDAT, nMSG** makes up the triplet.  Those values correspond to what you will find in the TWAIN specification under the "Twain Triplets" umbrella.  Recording those values will indicate what triplet was being sent.

**Comment by dynarithmic** [comment]
As to the screenshot, you could try and get the MS symbols by checking the "Microsoft Symbol Server".   This should load the symbols for ntdll.dll, and then you can view the call stack to see what function is being called on the OS side.

Also, since you were successful in getting the dtwain64u.dll to compile, try to compile dtwain64ud.dll (the debug dll).  You will have to change your C# code to use DTWAIN64UD.CS and rebuild.  The debug DLL's are not optimized, so the source code will match up with the actual function being called.

**Comment by sean-neeley** [comment]
Thanks for the tips.  The main thread is indeed stuck on that line of code in ctlwmgr.cpp.  I used your advice and ran with DTWAIN64UD.DLL.  I was able to see the values passed in:

![image](https://user-images.githubusercontent.com/80922869/219224255-72ce9fbc-d03a-4fb7-b92b-9c3966ee3b1e.png)

**Comment by sean-neeley** [comment]
Call stack:


ntdll.dll!NtMapViewOfSection ()
ntdll.dll!LdrpMapViewOfSection()
ntdll.dll!LdrpMinimalMapModule()
ntdll.dll!LdrpMapDllWithSectionHandle()
ntdll.dll!LdrpMapDllNtFileName()
ntdll.dll!LdrpMapDllSearchPath()
ntdll.dll!LdrpProcessWork ()
ntdll.dll!LdrpLoadDllInternal()
ntdll.dll!LdrpLoadDll ()
ntdll.dll!LdrLoadDll()
KernelBase.dll!LoadLibraryExW()
GdiPlus.dll!CreateWicImagingFactory(struct IWICImagingFactory * *)
GdiPlus.dll!GpWicCodec<class GpWicBmpEncoder,class GpWicBmpDecoder>::GetWicFactory(void)
GdiPlus.dll!GpWicDecoder::InitDecoder()
GdiPlus.dll!CreateDecoderForStream()
GdiPlus.dll!GpDecodedImage::GpDecodedImage(struct IStream *)
GdiPlus.dll!GpDecodedImage::CreateFromStream()
GdiPlus.dll!GpDecodedImage::CreateFromFile(unsigned short const *,class GpDecodedImage * *)
GdiPlus.dll!CopyOnWriteBitmap::DereferenceStream()
GdiPlus.dll!CopyOnWriteBitmap::CopyOnWriteBitmap()
GdiPlus.dll!GpBitmap::GpBitmap(unsigned short const *)
GdiPlus.dll!GdipCreateBitmapFromFile()
hpxsTwain.ds!00007fff59ed0e2e()
hpxsTwain.ds!00007fff59ed0cb7()
hpxsTwain.ds!00007fff59d6999f()
hpxsTwain.ds!00007fff59d69e65()
hpxsTwain.ds!00007fff59d7673d()
hpxsTwain.ds!00007fff59d74b7a()
hpxsTwain.ds!00007fff59bca0dc()
TWAINDSM.dll!000000018000491a()
TWAINDSM.dll!00000001800095eb()
dtwain64ud.dll!dynarithmic::CTL_TwainAppMgr::CallDSMEntryProc(const dynarithmic::CTL_TwainTriplet & pTriplet) Line 2658
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltwmgr.cpp(2658)
dtwain64ud.dll!dynarithmic::CTL_TwainTriplet::Execute() Line 98
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltrp.cpp(98)
dtwain64ud.dll!dynarithmic::CTL_ImageXferTriplet::Execute() Line 83
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltr026.cpp(83)
dtwain64ud.dll!dynarithmic::CTL_ImageMemXferTriplet::Execute() Line 158
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltr031.cpp(158)
dtwain64ud.dll!dynarithmic::CTL_TwainAppMgr::StartTransfer(dynarithmic::CTL_ITwainSession * __formal, dynarithmic::CTL_ITwainSource * pSource, dynarithmic::CTL_ImageXferTriplet * pTrip) Line 1447
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltwmgr.cpp(1447)
dtwain64ud.dll!dynarithmic::CTL_TwainAppMgr::BufferTransfer(dynarithmic::CTL_ITwainSession * pSession, dynarithmic::CTL_ITwainSource * pSource, bool bIsMemoryFile) Line 1192
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltwmgr.cpp(1192)
dtwain64ud.dll!dynarithmic::CTL_TwainAppMgr::TransferImage(const dynarithmic::CTL_ITwainSource * pSource, int nImageNum) Line 954
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltwmgr.cpp(954)
dtwain64ud.dll!dynarithmic::CTL_ProcessEventTriplet::ExecuteEventHandler() Line 171
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltr008.cpp(171)
dtwain64ud.dll!dynarithmic::CTL_TwainAppMgr::IsTwainMsg(tagMSG * pMsg, bool bFromUserQueue) Line 843
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltwmgr.cpp(843)
dtwain64ud.dll!DTWAIN_IsTwainMsg(tagMSG * pMsg) Line 321
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctldtwaininit.cpp(321)
dtwain64ud.dll!dynarithmic::TwainMessageLoopV2::CanEnterDispatch(tagMSG * pMsg) Line 519
	at C:\Users\SEN\Projects\dtwain5_library\h\ctliface.h(519)
dtwain64ud.dll!dynarithmic::TwainMessageLoopWindowsImpl::PerformMessageLoop(dynarithmic::CTL_ITwainSource * pSource, bool isUIOnly) Line 56
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctltwainmsgloop.cpp(56)
dtwain64ud.dll!dynarithmic::SourceAcquireWorkerThread(dynarithmic::SourceAcquireOptions & opts) Line 407
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctlsourceacquire.cpp(407)
dtwain64ud.dll!dynarithmic::SourceAcquire(dynarithmic::SourceAcquireOptions & opts) Line 250
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctlsourceacquire.cpp(250)
dtwain64ud.dll!dynarithmic::AcquireExHelper(dynarithmic::SourceAcquireOptions & opts) Line 424
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctlsourceacquire.cpp(424)
dtwain64ud.dll!DTWAIN_AcquireBufferedEx(void * Source, long PixelType, long nMaxPages, long bShowUI, long bCloseSource, void * Acquisitions, long * pStatus) Line 41
	at C:\Users\SEN\Projects\dtwain5_library\cpp\ctlacquirebuffered.cpp(41)
mscorwks.dll!DoNDirectCall__PatchGetThreadCall ()
00007ff7a67a4e29()
00007ff7a67a3e05()
System.Windows.Forms.ni.dll!00007fff62a88f0e()
System.Windows.Forms.ni.dll!00007fff6292d8e7()
System.Windows.Forms.ni.dll!00007fff62d93870()
System.Windows.Forms.ni.dll!00007fff621e528d()
System.Windows.Forms.ni.dll!00007fff621e4fa6()
System.Windows.Forms.ni.dll!00007fff621e4e45()
System.Windows.Forms.ni.dll!00007fff621e3744()
mscorwks.dll!UMThunkStubAMD64 ()
user32.dll!UserCallWinProcCheckWow()
user32.dll!DispatchMessageWorker()
mscorwks.dll!DoNDirectCall__PatchGetThreadCall ()
System.Windows.Forms.ni.dll!00007fff62200db7()
System.Windows.Forms.ni.dll!00007fff621ffd63()
System.Windows.Forms.ni.dll!00007fff621ff5d8()
System.Windows.Forms.ni.dll!00007fff621ff045()
00007ff7a67a0163()
mscorwks.dll!CallDescrWorker ()
mscorwks.dll!CallDescrWorkerWithHandler ()
mscorwks.dll!MethodDesc::CallDescr(unsigned char const *,class MetaSig *,unsigned __int64 const *,int,int)
mscorwks.dll!ClassLoader::RunMain(class MethodDesc *,short,int *,class PtrArray * *)
mscorwks.dll!Assembly::ExecuteMainMethod(class PtrArray * *)
mscorwks.dll!SystemDomain::ExecuteMainMethod(struct HINSTANCE__ *,unsigned short *)
mscorwks.dll!ExecuteEXE(struct HINSTANCE__ *)
mscorwks.dll!_CorExeMain ()
mscoreei.dll!_CorExeMain ()
mscoree.dll!_CorExeMain_Exported ()
kernel32.dll!BaseThreadInitThunk()
ntdll.dll!RtlUserThreadStart ()

**Comment by dynarithmic** [comment]
1) From the call stack, the triplet **(DG/DAT/MSG)** that is being sent from DTWAIN to TWAIN is

DG_IMAGE / DAT_IMAGEMEMXFER / MSG_GET = 2 / 259 / 1
This initiates the buffered transfer by sending this to TWAINDSM.DLL.  

2) From there, TWAINDSM.DLL calls the HP driver (hpxsTwain.ds) dll to continue processing this triplet.  

3) HP's driver code then calls GDI+, which is the imaging component present in all modern Windows systems.  The function is **GdiPlus.dll!GdipCreateBitmapFromFile()**

4) GDI+ takes over and calls **GdiPlus.dll!CreateWicImagingFactory(struct IWICImagingFactory * *)**, which according to [this link](https://learn.microsoft.com/en-us/windows/win32/api/wincodec/nn-wincodec-iwicimagingfactory) creates encoders, decoders and converters for the image.

5) GDI+ then calls the OS function **LoadLibraryExW()**, which dynamically loads the DLL.  It seems to have succeeded loading the library, but then is stuck on [NtMapViewOfSection](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-zwmapviewofsection).

**Comment by dynarithmic** [comment]
So this has something to do with the HP driver calling out to GDI+ in an attempt to process the image being scanned internally.

What if you changed the C# application to try native mode (DTWAIN_AcquireNativeEx), using the source UI?  This should be easier to see if there is an issue with the memory transfer and native transfer, since native transfer requires no memory to allocate on the DTWAIN side for the image.

I know that for no-UI, you're basically on your own trusting the driver to do everything correctly, unlike with the source UI being shown, where all the buttons, widgets, etc. are set up correctly via the user interface, and the driver simply reads those values off from those widgets and controls.

If the source UI run works, then the issue is most likely one of these scenarios:

1) using no UI, the HP driver needs to be initialized with certain capabilities before retrieving the image, but we don't know which ones to set.  In no-UI mode we are trusting that everything on the HP side will handle any "bad" settings it may have encountered.

2) The HP driver may have a bug in the no-UI mode.  This wouldn't be the first time I've seen drivers from other companies behave erratically when the UI is turned off.

3) Something lower level in the OS is causing the issue, mainly GDI+ trying to initialize itself by calling the OS kernel functions.

Since there is no way to get the HP's driver code to debug that, then these are the conclusions I can come up with right now.

**Comment by dynarithmic** [comment]
Also, I am running Windows 10, and it seems you are running Windows 11.  Is it possible to run this under Windows 10 on your side?  I have not run into any issues with Windows 10.

**Comment by sean-neeley** [comment]
Ok, I switched the code to use AcquiredNativeEx and turned UI mode on.  The application still hung.  This is the hpxsTwain.ds thread call stack when I pause:
![image](https://user-images.githubusercontent.com/80922869/219416115-36bff42c-75d4-46f8-8001-10a8dbc0d857.png)

And the main thread is waiting on the message loop.
![image](https://user-images.githubusercontent.com/80922869/219425501-72ce30da-4b01-4a46-96b1-e80d604f6f93.png)

I commented out the code in the C# test app for the callback as well, and that did not help.

I'll see if I can find a Windows 10 system.

**Comment by dynarithmic** [comment]
Try the DTWDEMO C++ application. It is different in that the message pump is within the application, and using DTWAIN_MODELESS processing, while the C# app uses the default DTWAIN_MODAL processing.

I am curious to see if the same scenario occurs with the stuck message pump in the C++ demo application.  If the C++ demo doesn't have any issues, then possibly this is a case where the C# code has to control the message loop, which I really was not hoping for.

**Comment by sean-neeley** [comment]
I think we tried that two weeks ago.  I just tried again with DTWDEMO64.EXE (C version) and it hangs when showing the Image Processing message.  I'm doing AcquireBufferedNative.  UI enabled and disabled does not make a difference. 
![image](https://user-images.githubusercontent.com/80922869/219470161-a76c2d23-c614-474c-887a-f6173123a10f.png)

If I run the 32bit DTWDEMO.EXE (C application), it completes without hanging when I test the same way.  
![image](https://user-images.githubusercontent.com/80922869/219470399-8903b85f-d9e8-40e7-a7f9-735d0156d9e6.png)

So I don't think this is a C# only issue.  In fact, we could continue troubleshooting without using C# if you want.
If you want me to try any code changes, I just need help with one compilation issue with DTWDEMO.EXE.  I get some kind of weird 9009 exit code when building:
![image](https://user-images.githubusercontent.com/80922869/219471550-07de1151-bc7b-4117-a44b-dce77d43e164.png)

Thanks

**Comment by sean-neeley** [comment]
Here is the stack trace when I pause the debugger while running the DTWDEMO64U.EXE (C application) when it is hung:
<img width="218" alt="image" src="https://user-images.githubusercontent.com/80922869/219475347-74499c5d-0611-45e9-b92c-249748ca2732.png">

HP driver thread:
<img width="175" alt="image" src="https://user-images.githubusercontent.com/80922869/219475609-91c14dc8-93c7-4e18-8ec7-028ea9a48756.png">

**Comment by dynarithmic** [comment]
1) You should run dtwdemo64ud.exe in the debugger, as that uses the debug DLL.  You will then see the line of code where this is hanging inside the demo application.

2) The errors are post-build step errors, and shouldn't affect the building of the actual executable itself.

**Comment by sean-neeley** [comment]
Here is the stack with DTWDEMO64UD.EXE:
<img width="517" alt="image" src="https://user-images.githubusercontent.com/80922869/219502342-e3757ed4-9196-46a4-bd8d-5a8de922fc0c.png">

I'm trying Windows 10 next...

**Comment by sean-neeley** [comment]
OK, I found a Windows 10 laptop.  Installed just the HP Basic driver and TWAINDSM.  Ran the DTWDEMO64.EXE on it.  AcquireNative.  It fed the paper through and then hung on the Processing Images dialog, just like Windows 11.  I also saw there was a November 2022 firmware update.  Tried that as well, but no luck.  Would my TWAINDSM library make any difference?  Which one are you using and where did you download it from?

**Comment by dynarithmic** [comment]
Here is the version of the DSM that's running:

![Clip1](https://user-images.githubusercontent.com/1283116/219513578-c5c203b2-9405-4481-a89c-c7fa033ace05.jpg)

I haven't changed this in a while.  You can usually get a version from [here](https://github.com/twain/twain-dsm/tree/master/Releases/dsm_020403/windows/64), but it seems old, and the source code is more up to date (you would have to build it yourself).

Just as a test, you can get the full path of the active DSM by calling DTWAIN_GetActiveDSMPathA or DTWAIN_GetActiveDSMPathW, after a connection to at least one source (or after the select source dialog is displayed).

**Comment by sean-neeley** [comment]
OK, we're using 2.5.0.0.  I'll roll back to your version and see if that makes a difference.  I can't think of anything else that would be different between our environments.

**Comment by dynarithmic** [comment]
Did you get the TWAINDSM.DLL from the installation of the HP 64-bit driver?  I may want to test the version you have.

As to the error, there is something odd going on with the message pump, where HP or the TWAIN DSM doesn't make progress and is stuck on the GetMessage() call.  I initially thought it would be that the HP dialog is doing something strange, but it could be the TWAINDSM.

Also, do you have another brand of scanner with a 64-bit driver that works?  I'm curious to see if the issue is for HP only or any 64-bit scanner.

**Comment by dynarithmic** [comment]
Here is the edited DTWAIN64UD.cs to add the pinvokes for DTWAIN_GetActiveDSMPath.
[dtwain64ud.zip](https://github.com/dynarithmic/twain_library-csharp_demo/files/10763363/dtwain64ud.zip)

**Comment by sean-neeley** [comment]
Thanks Paul.  We were having trouble with selecting the scanner source and were using TWAINDSM.DLL 2.1.2.0.  You suggested that we try the 2.4.0.0 version and gave us the link.  This post:
[[](https://github.com/dynarithmic/twain_library/issues/33#issuecomment-1162798479)](https://github.com/dynarithmic/twain_library/issues/33#issuecomment-1162798479)

I'm not sure how we got up to 2.5.0.0.   In the end, the problem may not have been with TWAINDSM.DLL, but I can't remember any more.   I may have compiled 2.5.0.0 from source.  I will test out some of the older versions and see what different it makes.   It may take me a few days to get to it.  I'll attach the 2.5.0.0 one that we have been using.

[TWAINDSM.dll.zip](https://github.com/dynarithmic/twain_library-csharp_demo/files/10768644/TWAINDSM.dll.zip)

The only other scanner we have is an old Fujitsu fi-6110.  I have the 32bit legacy twain driver installed for it.  I'll have to check the Fujitsu site for a 64bit twain driver.

**Comment by dynarithmic** [comment]
From the behavior of what you've described, I believe that the 64-bit HP driver requires the message loop to be active before the source UI attempts to show the user interface or if no user interface, right at the instance when a user interface would have been shown (but is suppressed).

There is a very little used function in DTWAIN:

`DTWAIN_ForceScanOnNoUI`

What this function does is post a message onto the message queue before the scanner is fully set up.  See if this has any effect.

The way you use it is as soon as you select a source, call the function:

`TwainAPI.DTWAIN_ForceScanOnNoUI(YourSource, 1)`

Also, scan without a user interface, as currently this function is only invoked on no UI.

**Comment by dynarithmic** [comment]
Also, here are my thoughts on why I believe TWACKER worked for you, and is related to the comment above.

TWACKER is a pure C, Windows GUI program.  The history of the TWAIN API is that it was solely meant for C language and GUI Windows applications.  Thus the API was based on the traditional Windows message pump to be active at all times.  

Since all GUI programs have a message pump, it was easy for C (and C++) Windows apps to adjust the message pump.  The problem is what to do with console apps that do not have such a mechanism, or languages where there is no access to the message pump.  This has always been the issue with the TWAIN API -- you had to write a C-based GUI program to make TWAIN work.  In the case of TWACKER, it has the messaging started even before a source has been selected.

What DTWAIN does is attempt to move the message pump inside the DLL, and only activate it when an acquisition is attempted.  So far, this has worked for all the scanners that I know of.  However the HP 64-bit scanner may be one of those rare ones that require the messaging to be on at all times.  Again, simple for a C/C++ GUI app, but maybe impossible for other languages.  So I am assuming that HP built their drivers for C/C++ applications.  

Note that this is an assumption, but to me this is by far the strongest case of behavior you're seeing.  So you should try the suggestion in the previous comment, while I have to think about how to get around the issue.  Note that I am not seeing the same issue on my end, so any changes are basically done blindly, hoping they will resolve the issue you're seeing.

**Comment by sean-neeley** [comment]
Thanks for all of the suggestions.  I will be trying these ideas on Tuesday and will report back.

**Comment by dynarithmic** [comment]
Just to keep you updated:

I made some coding changes over the past couple of days that will move the start of the TWAIN loop right before the showing of the user interface (or right before the user interface would have been shown if using no-UI).   Hopefully this helps in resolving the issue you're having with the 64-bit driver.  There isn't a way for me to confirm these changes will have an affect, since I haven't been able to duplicate your issue.

I did some rudimentary testing, and the scanning still works for the sample TWAIN driver.  I tested both native and buffered transfers.  I haven't tested file transfers, but they should still work.   I haven't tested what happens in scenarios where the acquisition legitimately fails, and to make sure the correct error return codes are produced.  That will take some time.

Any new DLL's will be placed in the [experimental-5.3.0.0](https://github.com/dynarithmic/twain_library_source/tree/experimental-5.3.0.0) branch.  I will let you know when the DLL's are ready.

**Comment by dynarithmic** [comment]
I've updated the [experimental branch](https://github.com/dynarithmic/twain_library_source/tree/experimental-5.3.0.0/binaries) with the new libraries.  I did a small file acquire test, and it looks like it works ok with the TWAIN loop adjustment.

Let me know if these address the issue.

**Comment by sean-neeley** [comment]
Hi Paul, I've got some good news -- I tried out your new dtwain64u.dll in both DTWDEMO64.EXE and our scanning application and the hanging is gone!  I'm still testing, but it looks promising.  AcquiredBuffereEx works great, but AcquiredNativeEx was crashing.  We don't use AcquireNative, but I'll try to help you figure out where it fails.

**Comment by sean-neeley** [comment]
FYI, I posted this to HP about the exceptions in their HP Driver.  The Driver is looking for a settings file in the wrong folder.
[https://h30434.www3.hp.com/t5/Scanning-Faxing-Copying/Exception-thrown-in-64bit-HP-TWAIN-driver-about-missing/m-p/8623506#M269310](https://h30434.www3.hp.com/t5/Scanning-Faxing-Copying/Exception-thrown-in-64bit-HP-TWAIN-driver-about-missing/m-p/8623506#M269310)

**Comment by dynarithmic** [comment]
👍  Great,  that sounds good.  

I uploaded new versions of DTWDEMO32 and DTWDEMO64 apps to the [experimental branch](https://github.com/dynarithmic/twain_library_source/tree/experimental-5.3.0.0/binaries).  These versions do not do modeless processing (note the commented out **DTWAIN_SetTwainMode(DTWAIN_MODELESS)** call), so will test the new modal loop processing internal to the DLL.  

The DTWAIN_AcquireNativeEx is the first one I tested, so I'm surprised it crashed.  If it continues to crash, you could build DTWAIN64UD.DLL, rebuild the demo app (the DTWDEMO project in the Visual Studio solution) and run it under the debugger.   The source code is based on the [experimental branch source](https://github.com/dynarithmic/twain_library_source/tree/experimental-5.3.0.0).  Hopefully that branch is up to date in terms of the necessary build steps (the source code should be up to date).

**Comment by sean-neeley** [comment]
I tried your new DTWDEMO64 version.  Works good for DTWAIN_AcquiredBufferedEx - no hanging.  I also tried DTWAIN_AcquireNativeEx and it did not crash this time.  However the app behaves differently with AcqurieNativeEx.  If I scan 4 pages, this popup appears only for the first page and the preview image is blank.

<img width="645" alt="image" src="https://user-images.githubusercontent.com/80922869/220438702-340d65f6-d140-43d3-ad51-f55229aaba15.png">

With AcquiredBufferedEx, it shows a preview for each page scanned asking if I want to keep it or not.

EDIT: I added a video of AcquireBufferedEx and AcquireNativeEx working differently when scanning the same 2 pages.
https://user-images.githubusercontent.com/80922869/220442295-bc7c148d-7c7f-43d3-8d95-42d2c71044df.mov

**Comment by dynarithmic** [comment]
ok, I will check it out and see what the issue is with the native mode.

**Comment by sean-neeley** [comment]
Thanks! We don't use native mode, so it is not a problem for us and no rush.  I just thought you should know about it before you turn this into a new release.

**Comment by dynarithmic** [comment]
I debugged further, and for some very odd reason, the native mode is returning an invalid bitmap from the scanner.  I tested with the TWAIN sample source, and the images delivered by the scanner are valid.   On the other hand, the pointer address that the HP driver is giving back, which is supposed to be the address of the image data, is invalid.

So I would go with just the buffered mode for now.  It is strange that the more complex mode, buffered, is working as it should, but native isn't, which is just asking the scanner to hand back a pointer to the image data after being scanned.

Also, the 32-bit HP driver works for both native and buffered.  I am wondering if this is indeed a bug in the HP 64-bit driver when it comes to native mode.

**Comment by dynarithmic** [comment]
OK.  I tested with TWACKER64, and the image comes back.  I will just have to debug further to see why the native transfer is not returning a valid bitmap.

**Comment by sean-neeley** [comment]
OK sounds good!

**Comment by dynarithmic** [comment]
Update:

I tested with 
1) the Plustek scanner 64-bit driver
2) The TWAIN Sample 64-bit driver
3) The HP-7000 S3

Both the Plustek and the TWAIN Sample produce valid bitmaps for 64-bit.  However, the HP driver does not return a valid bitmap handle after requesting the bitmap that was scanned.

Technical Details:
I verified this by closely watching what happens on this call to TWAIN:

DG_IMAGE / DAT_IMAGENATIVEXFER / MSG_GET

This triplet is the one that tells Twain and the HP driver to "give me the bitmap you just scanned in native mode".   You will find the call of this triplet in ctltwmgr.cpp, around line 1038:

```cpp
int CTL_TwainAppMgr::NativeTransfer( CTL_ITwainSession *pSession,
                                      CTL_ITwainSource  *pSource)
{
   CTL_ImageXferTriplet IXfer(pSession, pSource, DAT_IMAGENATIVEXFER);
    return StartTransfer( pSession, pSource, &IXfer );  // <-- Here is where we ask to get the bitmap that was just scanned
}
```
The scanning process has already been completed before this triplet is sent out -- so the scanner has produced the image already.  

On input, we give TWAIN a handle to fill in with the bitmap handle that was scanned.   The HP scanner **does** scan correctly -- if you scan with the user-interface, you see the HP Twain dialog show the scanned image in the HP's application.  However when DTWAIN asks for the bitmap when issuing the triplet call I mentioned, HP returns *something*, but whatever it is, it isn't a valid bitmap.  As a matter of fact, it is a pointer to nowhere valid.

That's why in the DTWDEMO application, you don't see an image -- the image handle returned by HP is not valid.

I may have to test another 64-bit scanner (Plustek, HP, and the TWAIN Sample source are what I have to go on right now, as all the other scanners I have only have a 32-bit driver).  If another scanner (maybe Fujitsu or Epson, 64-bit) works correctly, then I will try and make a case to HP that they should take a look at their native transfer 64-bit driver code.  It is certainly not behaving correctly when it is set in a certain state.   If you have another 64-bit scanner, then that would also be helpful.

The buffered transfer for HP works, since the setting up of the bitmap is done on the DTWAIN side.  All HP does is fill in our bitmap area with strips of data.  This is different than native mode, where HP controls everything, including creating the area for the bitmap.

As to TWACKER seemingly working -- the TWACKER program does minimal work to get the scan to go through -- possibly DTWAIN puts an extra burden on the driver (more capabilities are tested and set) that wasn't tested out by HP thouroughly (at least that's my guess).  There should be no reason why after a successful scan, HP returns an invalid bitmap back to DTWAIN.

The frustrating thing is that the source code to the HP driver is unavailable.  If I had access to it, it would take a very short time in getting to the cause of the issue.  The only part of this entire process that is unavailable is the HP driver code.  DTWAIN is open source, the TWAIN DSM is open source, and that's where the dead-end hits, and that is when it's calling the HP driver code.

So I will keep looking and testing for now.

**Comment by sean-neeley** [comment]
OK, I only have one other scanner, which is a Fujitsu.  I downloaded the legacy 32bit TWAIN driver for it and that is what I have been using.  There is a PaperStream IP 64bit driver that I can try to download and test.  

I registered for an HP developer account, but I don't think they provide the HP driver source.  They do give source code examples for using their SDK.   I can share a zip of it tomorrow if you are interested.

I have another question for you.  In the latest DLL did you recently add these callback messages?

<img width="395" alt="image" src="https://user-images.githubusercontent.com/80922869/220768780-ad091864-452b-40fa-82ca-ea07f64131e3.png">

I noticed our own application freezes up and acts weird now if you try to scan and there is no paper in the feeder.  I haven't tested with DTWDEMO64.EXE.  Did you modify anything related to the "no paper in the feeder" situation?  I'm not seeing that new DTWAIN_TN_FEEDERNOTLOADED come through in the callback function.  

These are some example errors we do see.  What is the best way to deal with errors like this in the callback?

2023-02-22 13:32:50,388 [1] INFO  PProScanning.MainForm - TwainCallbackProc wParam=DTWAIN_TN_GENERALERROR
2023-02-22 13:32:50,388 [1] INFO  PProScanning.MainForm - Error -1118: TWAIN Condition Code: Feeder is jammed


2023-02-22 13:32:50,388 [1] INFO  PProScanning.MainForm - TwainCallbackProc wParam=DTWAIN_TN_GENERALERROR
2023-02-22 13:32:50,388 [1] INFO  PProScanning.MainForm - Error -1023: TWAIN user interface could not be started.  Device error or Acquire must have a user interface.

2023-02-22 13:41:12,224 [1] INFO  PProScanning.MainForm - TwainCallbackProc wParam=DTWAIN_TN_GENERALERROR
2023-02-22 13:41:12,224 [1] INFO  PProScanning.MainForm - Error -1128: TWAIN Condition Code: Operation failed (source has nothing to capture for transfer)


Thanks!

**Comment by dynarithmic** [comment]
The feeder notifications are not used, and reserved for future use.  It was easier to determine if paper was in the feeder, it is left up to the application.  You would do this by polling the DTWAIN_IsFeederLoaded() in a loop, and breaking out of the polling loop once DTWAIN_IsFeederLoaded() returns TRUE.  (You could also set up a timer that breaks out of the loop if the time expires -- that's what is done in the latest C++ wrapper for DTWAIN).

As far as modifying the "no paper in the feeder", the message loop was moved, so I wouldn't be surprised if some behavior has changed.  But again, determining if paper is in the feeder is best done by polling, in my opinion.
The TRIPLET notifications are relatively new, and just notifies the application of the triplets being sent to the TWAIN system.

**Comment by dynarithmic** [comment]
As to handling the notification errors, you would probably want to trap the error, inform the user of what happened (like a paper jam), and then close/reopen the source.

**Comment by sean-neeley** [comment]
OK, I did some more testing today.  When using DTWDEMO64U.exe and the new DLL, if the document feeder is empty and you try to scan with the HP 7000 S3, the HP "scanning" dialog pops open for a few milliseconds then closes.  Control returns to the application and you can then put paper in the feeder and try again and all is good.

However when you use the C# FullDemo, in 64bit mode and the same dtwain64u.dll, the application has problems when you leave the document feeder empty and try to scan.  AcquireBufferedEx hangs forever.  I tried putting a TwainAPI.DTWAIN_IsFeederLoaded(SelectedSource) call right before the AcquireBufferedEx call.  It always returns 1, even if there is no paper in the feeder.  If I try the C# FullDemo with the old 32bit dtwain32u.dll, TwainAPI.DTWAIN_IsFeederLoaded(SelectedSource) correctly indicates that the feeder is empty.

Is this something you can look into?  Whenever someone forgets to load papers into the document feeder, our application hangs with the new dtwain64u.dll and I don't have a work-around.

Also FYI, I could not find a 64bit twain library that would work with our old Fujitsu fi-6110 scanner.  Therefore the only scanner I can do any 64bit twain library testing on is the HP scanjet 7000 s3.

**Comment by dynarithmic** [comment]
For 64-bit, if there is no paper in the feeder, and you called the DTWAIN_IsFeederLoaded() function, and it returns TRUE, then that is more than likely a bug in the HP driver.   

To verify, I ran a simple application that opens the HP source, calls DTWAIN_IsFeederLoaded() with the feeder empty-- indeed the HP driver returns a 2, which is absolutely wrong.

It is wrong in two ways -- 1) There is no paper in the feeder (of course), but more importantly, the capability (CAP_FEEDERLOADED) is a boolean capability -- it is only supposed to return TRUE (1) or FALSE (0).  This is clearly outlined in the TWAIN specficiation.  The value of 2 is not defined for TRUE or FALSE, thus the default behavior for DTWAIN_IsFeederLoaded() is to treat this value as TRUE.

When I run the 32-bit HP driver, it correctly returns TRUE or FALSE, depending on whether there is paper in the feeder or not.

If you have the TWAIN spec, go to the CAP_FEEDERLOADED description.  Here is a synopsis:

```
CAP_FEEDERLOADED

Description
Reflect whether there are documents loaded in the Source’s feeder.
This is a read only capability.

Application
Used by application to inquire whether there are documents loaded in the Source’s feeder.

Source
If CAP_FEEDERENABLED equals FALSE, return TWRC_FAILURE / TWCC_CAPSEQERROR
(capability is not supported in current settings).

If CAP_FEEDERENABLED equals TRUE, return the status of the feeder (documents loaded = no documents loaded = FALSE).
...
Values Type:
TW_BOOL

Allowed Values:
TRUE or FALSE
```
So it looks like HP is not following the spec and/or their driver code for the 64-bit version is broken. 

You could verify this in the C# code (maybe create a very small program) that opens the HP source, and either calls DTWAIN_IsFeederLoaded(), or try 

`DTWAIN_GetCapValues(Source, CAP_FEEDERLOADED, DTWAIN_GET, YourArray) 
`
and inspect the values from YourArray by calling 

`DTWAIN_ArrayGetAtLong(YourArray, 0, ref int); 
`
It's a little lower level, but you're removing one level of abstraction when talking to the TWAIN system.  DTWAIN_IsFeederLoaded basically does the same thing, but in a more convenient way for the end-user.

**Comment by dynarithmic** [comment]
I just tested the Plustek scanner for 64-bit.  The Plustek does the correct reporting of paper in the feeder.  TRUE is returned when paper is in the feeder, FALSE otherwise.  The Plustek scanner I am using is model PS286 Plus.  So if you're in the business of recommending what scanner to use for 64-bit processing, this model seems to hold up rather well.

**Comment by sean-neeley** [comment]
Strange.  I wrote those lines of code and tested various capabilities.  I get back the value 2 for several things: CAP_FEEDERLOADED, CAP_FEEDERENABLED and CAP_DEVICEONLINE.  Maybe others.  I'll have to mess with it some more when I get a chance.  Any hope of HP fixing something like this?  What is the best way to bring it to their attention?  The issue we have is that we have probably 100 of these HP scanners out in the wild and they can't be replaced.  I'm starting to think we will have to give up on the 64bit driver and just build and release a 32bit application that uses the 32bit driver.

**Comment by dynarithmic** [comment]
I've never contacted HP myself, so not sure how to get a fast response (if any response) from them.  

I just tested the 64-bit TWACKER.  I made a change in the source code to check CAP_FEEDERLOADED.  It is returning a value of 2, just like the DTWAIN Demo program.  So this confirms what is being seen, since TWACKER is totally independent of DTWAIN in every way, as it is source code directly from the Twain Working Group website.

If you are lucky to bring it to HP's attention, they may want to take a look at their 64-bit driver code -- maybe also bring up other scanner drivers and manufacturers that do the correct thing using the same TWAIN-enabled programs, and more importantly their driver fails TWACKER.

So far, the message loop scenario could be a DTWAIN thing, and it seems to have been addressed.  However the other issues, such as native mode returning an invalid bitmap, and the obvious errors in their reporting on the state of the scanner in 64-bit mode, HP needs to address those.

As to your application, you *could* still have it work for all installed scanners that are not HP 7000, but filter out the selection in the Select Source dialog.  That way, users who have "good" scanners still get the 64-bit app, and users with HP scanners (at least the 7000 series), well, you have an explanation of why it was filtered out (buggy driver).  This way, there would be no way for the user to select the HP driver, thus no surprises and apps that behave erratically.

**Comment by dynarithmic** [comment]
If HP has a developer forum for their TWAIN applications, maybe you can create a forum item to bring up the behavior of their driver with respect to the TWACKER application.  

HP would be more likely to get their driver working correctly with TWACKER, since basically all TWAIN developers use it to some extent.

**Comment by sean-neeley** [comment]
Thanks, I submitted the following report to HP today on their developer portal, although I don't know if this is the right way to get their attention:

> Issue #1: When testing the 64bit driver using a C# program and opening the scanner source, exceptions occur because the HP driver is looking in the wrong folder for a Factory Settings file named PdsmqLocalSettings.xml.  This also causes breakpoints to trigger when running the application from Microsoft Visual Studio.  There are more details about this problem here:
> https://h30434.www3.hp.com/t5/Scanning-Faxing-Copying/Exception-thrown-in-64bit-HP-TWAIN-driver-about-missing/m-p/8623506
> 
> Issue #2: The HP driver does not return a valid handle to a bitmap after scanning using the native transfer mode.  This is the TWAIN triplet:
> DG_IMAGE / DAT_IMAGENATIVEXFER / MSG_GET
> The scanner scans the documents, but the handle returned is not valid and cannot be used to obtain the scanned image in native transfer mode.
> 
> Issue #3: In the TWAIN spec, CAP_FEEDERENABLED and CAP_FEEDERLOADED can be used to check if the scanner has a document feeder and if the feeder has paper.  According to the TWAIN specification, the driver should always return a 0 or 1 value.  When testing, the HP 64bit driver returns a value of 2, which is not valid.  It returns 2, regardless of whether there is paper in the feeder or not.  This causes serious issues in our application because the application thinks there is paper to scan when there is not paper in the document feeder.  I have noticed other checks, like CAP_DEVICEONLINE return an invalid value of 2 as well, so this is probably a widespread bug in the driver.  We tested this by modifying the 64bit TWACKER application.

**Comment by dynarithmic** [comment]
Thanks.  That looks ok.  It basically covers all of the issues seen so far.  Someone reading the forum posts at HP should have some sort of contact with engineering in some way, I would believe.

I think issue #3 is one where HP can see for themselves right away if they use TWACKER.   

For issue #2, they may have to use DTWAIN in some way to duplicate the issue.  More than likely, DTWAIN puts the driver in a state where their code goes off the rails in returning a proper bitmap through to the TWAINDSM, and eventually DTWAIN.   

As to issue #2, have you considered distributing the 64-bit app, but simply filtering out the "HP TWAIN USB" selection from one of the sources to select?

**Comment by dynarithmic** [comment]
I took another look at the TWAIN specification, and the container types have changed between the version 1.x DSM and 2.x DSM, thus will require a fix to DTWAIN for the feeder issues.  

As a workaround, can you make the call in your C# code to test if the feeder is loaded by issuing a DTWAIN_GetCapValues (basically what you have done already), specifying that the cap to use is CAP_FEEDERLOADED, with the one difference being DTWAIN_CAPGETCURRENT instead of DTWAIN_CAPGET, as the cap operation?  

In the past, for the feeder capabilities, DTWAIN_CAPGET and DTWAIN_CAPGETCURRENT did the same thing, but it seems that 2.x of the DSM made it an option for scanners to make these operations return different information.   

Let me know if this solves the issue -- in the meantime, I will get new DTWAIN DLL's out to the experimental branch.   I will have to check the other caps such as CAP_DEVICEONLINE to make sure it uses DTWAIN_CAPGETCURRENT as the operation, and not the DSM version 1.x DTWAIN_CAPGET.

As to the invalid bitmap on native transfers, that will be the one outstanding issue, plus the issue of the HP driver throwing an exception on opening and closing the source, due to the missing XML file.

**Comment by dynarithmic** [comment]
Please try the latest DLLs on the [experimental branch](https://github.com/dynarithmic/twain_library_source/tree/experimental-5.3.0.0/binaries).  

The DLL's should address the issues you're having with the feeder status, and hopefully you can get your application to work correctly (not withstanding the outstanding issues of native acquisition and the HP exceptions being thrown on opening of the source).

**Comment by sean-neeley** [comment]
This sounds very promising!  Great find!   I'll be able to test it out tomorrow.

**Comment by sean-neeley** [comment]
Ok, I tested and your changes allow us to get the correct feeder status!  This is great, and we now have a fully working application with the HP 64bit driver.

**Comment by dynarithmic** [comment]
That sounds good.  The new DLL's will probably be released in the next few days to the master branch.  Still going through some preliminary testing.

**Comment by sean-neeley** [comment]
Thanks, let me know when the new release is out and I will grab it.

**Comment by dynarithmic** [comment]
The [new release is ready](https://github.com/dynarithmic/twain_library).

**Comment by Jayeshshetye2025** [comment]
Hi , I have downloaded full demo zip file. Running properly. I clicked on Select Source By Name and enter Scanner name and then I clicked on Acquire Native. Scanning worked fine. Can we achieve same thing in One click?  Within one click user can enter Scanner Name to get selected source and then scanning will also happen. I tried that but TwainAPI.DTWAIN_AcquireNativeEx throws exception "TWAIN Condition Code: TWAIN triplet is out of expected sequence". Do i need to do some change in code ?? Please suggest
